<a href="https://colab.research.google.com/github/AntonioMoradoRamos/meia-masterdegree-experiments/blob/main/tarefas_ml_1_primeira_observacao_dados.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Primeira Observação dos Dados

Ficheiro "data1.xlsx" (Full Year of historical data (1st data set) (53 907 KB) ).<br>
Importar o ficheiro para o colab.<br>

O dataset "ds_building_energy" corresponde à folha "building_energy"<br>
O dataset "ds_weather_data" corresponde à folha "weather_data"<br>



##Bibliotecas

In [ ]:
import pandas as pd

from datetime import date

In [ ]:
# Versão do python
!python --version

Python 3.7.12


In [ ]:
# Versão do panas
print(pd.__version__)

1.1.5


##Carregar os conjuntos de dados

In [ ]:
def load_excel_data():
  return pd.ExcelFile('data1.xlsx')

In [ ]:
xsls = load_excel_data()

In [ ]:
ds_building_energy = pd.read_excel(xsls, 'building_energy')
ds_weather_data = pd.read_excel(xsls, 'weather_data')

## Análise exploratória

In [ ]:
ds_building_energy.head()

date      time  consumption (w)  generation (w)
0 2019-01-01  00:05:00             2985               0
1 2019-01-01  00:10:00             2258               0
2 2019-01-01  00:15:00             2266               0
3 2019-01-01  00:20:00             3016               0
4 2019-01-01  00:25:00             2265               0

In [ ]:
ds_weather_data.head()

datetime (epoch)  temperature (˚C)  humidity (%)  radiation (Wm^2)
0        1546301100               8.7          76.0               0.0
1        1546301400               8.6          76.0               0.0
2        1546301700               8.7          76.0               0.0
3        1546302000               8.7          76.0               0.0
4        1546302300               8.6          76.0               0.0

In [ ]:
ds_building_energy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105120 entries, 0 to 105119
Data columns (total 4 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   date             105120 non-null  datetime64[ns]
 1   time             105120 non-null  object        
 2   consumption (w)  105120 non-null  int64         
 3   generation (w)   105120 non-null  int64         
dtypes: datetime64[ns](1), int64(2), object(1)
memory usage: 3.2+ MB


In [ ]:
ds_weather_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 97191 entries, 0 to 97190
Data columns (total 4 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   datetime (epoch)  97191 non-null  int64  
 1   temperature (˚C)  95387 non-null  float64
 2   humidity (%)      95364 non-null  float64
 3   radiation (Wm^2)  95092 non-null  float64
dtypes: float64(3), int64(1)
memory usage: 3.0 MB


### Observação dos registo do mês de janeiro

In [ ]:
def convert_int64_to_timestamp(dateAsInt64):
  #print("datetime as int64: " , dateAsInt64)
  theTimeStamp = pd.to_datetime(dateAsInt64, unit='s')
  #print("datetime as timestamp: " , theTimeStamp)

  return theTimeStamp

In [ ]:
def split_timestamp_to_date_and_time(ds, date_time_feature_name):
  series_date =  pd.to_datetime(convert_int64_to_timestamp(ds[date_time_feature_name]))
  # Para além de obter a data, i.e., o ano, o mês e o dia, é preciso garantir que o tipo de dados é datetime64
  # Este é o tipo de dados dos outros conjuntos de dados. 
  # É preciso que seja igual para realizar o merge
  ds['date'] = pd.to_datetime(pd.to_datetime(series_date).dt.date) 

  series_time =  pd.to_datetime(convert_int64_to_timestamp(ds[date_time_feature_name]))
  ds['time'] = pd.to_datetime(series_time).dt.time

  return ds

In [ ]:
def ds_janeiro(ds):
  query = "date >= 20190101 and date <= 20190131"
  return ds.query(query);

In [ ]:
ds_building_energy_janeiro = ds_janeiro(ds_building_energy)

In [ ]:
ds_weather_data_date_and_time = split_timestamp_to_date_and_time(ds_weather_data, "datetime (epoch)")
ds_weather_data_date_and_time_janeiro = ds_janeiro(ds_weather_data_date_and_time)

In [ ]:
ds_building_energy_janeiro.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8927 entries, 0 to 8926
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   date             8927 non-null   datetime64[ns]
 1   time             8927 non-null   object        
 2   consumption (w)  8927 non-null   int64         
 3   generation (w)   8927 non-null   int64         
dtypes: datetime64[ns](1), int64(2), object(1)
memory usage: 348.7+ KB


In [ ]:
ds_weather_data_date_and_time_janeiro.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8927 entries, 0 to 8926
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   datetime (epoch)  8927 non-null   int64         
 1   temperature (˚C)  8927 non-null   float64       
 2   humidity (%)      8927 non-null   float64       
 3   radiation (Wm^2)  8927 non-null   float64       
 4   date              8927 non-null   datetime64[ns]
 5   time              8927 non-null   object        
dtypes: datetime64[ns](1), float64(3), int64(1), object(1)
memory usage: 488.2+ KB


## Conclusões



1.   O conjunto de dados "ds_weather_data" não apresenta a data dividida em dois atributos "date" e "time" como no conjunto de dados "ds_building_energy"

> *    Será preciso uniformizar os atributos. Optou-se por ter os dois atributos "date" e "time". Desta forma será mais fácil distinguir unidades temporais como o dia e a hora do registo.<br>
*    A união entre os dois conjuntos de dados será realizada com os atributos "date" e "time"
*    Os tipos de dados dos atributos "date" e "time" terão de ser igual nos dois conjuntos.


2.   O conjunto de dados "ds_building_energy" apresenta 105120 registos e o "ds_weather_data" apresenta 97191. Portanto, o segundo conjunto de dados contém menos registos.

> *    Ter em atenção na operação de merge para que não existam registos sem dados.

3.   Nos conjuntos de dados "ds_building_energy" e "ds_weather_data" todos os registos contém dados, pois estes apresentam o valor "non-null"

4.   Análise do mês de janeiro

> *   Nos dois conjuntos de dados todos os registos contém dados, pois estes apresentam o valor "non-null"
*   Os dois conjuntos de dados obtidos apresentam o mesmo número de registo de 8927 entradas




#Bibliografia

*   Dataset gecad - smartgridcompetitions<br>
http://www.gecad.isep.ipp.pt/smartgridcompetitions/data/

*   Python<br>
https://docs.python.org/3.7/library/datetime.html#module-datetime<br>
https://docs.python.org/3.7/library/datetime.html#datetime.datetime<br>
https://docs.python.org/3/library/datetime.html


*   Pandas<br>
https://stackoverflow.com/questions/26521266/using-pandas-to-pd-read-excel-for-multiple-worksheets-of-the-same-workbook<br>
https://pandas.pydata.org/docs/reference/api/pandas.read_excel.html<br>
https://pandas.pydata.org/docs/getting_started/index.html#getting-started<br>
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.to_datetime.html<br>
https://pandas.pydata.org/docs/reference/api/pandas.Series.dt.date.html?highlight=date#pandas.Series.dt.date<br>
https://pandas.pydata.org/docs/reference/api/pandas.Series.apply.html

